openstack/utils/templates/_hosts.tpl

```diff
@@ -102,12 +161,70 @@ mysql+pymysql://{{ include "db_credentials" . }}@
         {{- end -}}
         charset=utf8
     {{- end }}
-{{- end}}
+{{- end }}
+
+{{/*
+Use .Values.pxc_db for connection URL generation
+*/}}
+{{- define "utils._db_url_pxc_db" }}
+    {{- if kindIs "map" . }}
+        {{- if and .Values.pxc_db.users .Values.pxc_db.databases }}
+            {{- $db := first .Values.pxc_db.databases }}
+            {{- $user := get .Values.pxc_db.users $db | required (printf ".Values.pxc_db.%v.name & .password are required (key comes from first database in .Values.pxc_db.databases)" $db) }}
+            {{- tuple . $db $user.name (required (printf "User with key %v requires password" $db) $user.password) | include "utils._db_url_pxc_db" }}
+        {{- else }}
+            {{- tuple . (coalesce .Values.dbName .Values.db_name) (coalesce .Values.dbUser .Values.global.dbUser "root" | include "resolve_secret_urlquery") (coalesce .Values.dbPassword .Values.global.dbPassword .Values.pxc_db.root_password | include "resolve_secret_urlquery" | required ".Values.pxc_db.root_password is required!") .Values.pxc_db.name | include "utils._db_url_pxc_db" }}
+        {{- end }}
+    {{- else -}}
+mysql+pymysql://{{ include "db_credentials" (dict "context" . "dbType" "pxc-db") }}@
+        {{- $allArgs := . }}
+        {{- $schemaName := index . 1 }}
+        {{- with $envAll := index . 0 }}
+            {{- if not .Values.proxysql }}
+                {{- include "utils._db_host_pxc_db" $allArgs }}
+            {{- else if not .Values.proxysql.mode }}
+                {{- include "utils._db_host_pxc_db" $allArgs }}
+            {{- else if ne $envAll.Values.proxysql.mode "unix_socket" }}
+                {{- if mustHas $envAll.Values.proxysql.mode (list "unix_socket" "host_alias") }}
+                {{- include "utils._db_host_pxc_db" $allArgs }}
+                {{- else }}
+                    {{ fail (printf "Unknown value for .Values.proxysql.mode: got \"%v\"" $envAll.Values.proxysql.mode) }}
+                {{- end }}
+            {{- end -}}
+            /{{ $schemaName }}?
+            {{- if .Values.proxysql }}
+                {{- if eq .Values.proxysql.mode "unix_socket" -}}
+                    unix_socket=/run/proxysql/mysql.sock&
+                {{- end }}
+            {{- end }}
+        {{- end -}}
+        charset=utf8
+    {{- end }}
+{{- end }}
 
 # Please keep as it is, special case when it has to reference the db_region value.
-{{define "db_host_pxc"}}{{.Release.Name}}-percona-pxc.{{.Release.Namespace}}.svc.kubernetes.{{.Values.global.db_region}}.{{.Values.global.tld}}{{end}}
+{{- define "db_host_pxc" }}
+{{- .Release.Name -}}-percona-pxc.{{- .Release.Namespace -}}.svc.kubernetes.{{- .Values.global.db_region -}}.{{- .Values.global.tld -}}
```

Comment:
No need for the hyphens, when there is no whitespace in between.
```
{{- .Release.Name }}-percona-pxc.{{ .Release.Namespace }}.svc.kubernetes.{{ .Values.global.db_region }}.{{ .Values.global.tld }}
```
The `{{- end }}` below eats the whitespaces before.

In general, I'd follow the pattern only to "eat" the whitespaces on the left-hand side, unless there is a good reason not to (i.e. whitespaces for readability) `{{- ...}}` .